Let the configure (rather than build) system handle the compiler's version

shindere opened the pull request:
With this PR, OCaml's version is defined in `build-aux/ocaml_version.m4`.

At first glance, it can seem the PR just replaces one dirty hack with
another but, with the new definition, all the version-related computations
while `configure` is generated rather than at each build.

At the moment, the PR keeps the `VERSION file`, which gets generated
at the same time than `configure`.

Since we don't install `VERSION` any longer, we could also stop generating
it completely, since it is (almost) no longer used.

One exception to this is `testsuite/tests/ast-invariants/test.ml` wich
checks for its presence, but I assume we could as well check for the
presence of another file?

There is also the case of `tools/check-parser-uptodate-or-warn.sh`, any
thought on what should be done there, @gasche?

The two other things this PR could change are:

 1. The release docs, @Octachron for directions

 2. The continuous integration, @dra27, also for a general review.

Also cc @damiendoligez who implemented the `VERSION` file, I think.

shindere committed:
Document where the OCaml version is defined and how to update the VERSION file

shindere committed:
Transform runtime/caml/version.h into a configured header

This means that, before this commit, runtime/caml/version.h was
generated during the build process by a recipe found in runtime/Makefile.

After this commit, runtime/caml/version.h is generated by the configure
script from runtime/caml/version.h.in

This commit makes an effort to guarantee that the version.h file
generated by configure is as similar as possible to the one produced
by the make recipe before, that's why runtime/caml/version.h.in
contains no copyright notice at this stage.
It will be added in the following commit.

At this stage, the only difference between the two versions of version.h
(the one generated by make before and the one generated by configure now)
is the coment line produced by configure as the first line of version.h, namely:

/* runtime/caml/version.h.  Generated from version.h.in by configure.  */

but this cannot be avoided and is harmless.

shindere committed:
Add copyright header to runtime/caml/version.h.in

shindere committed:
Introduce the OCAML_VERSION_EXTRA C preprocessor macro

This is a shorter alias for OCAML_VERSION_ADDITIONAL

shindere committed:
ocamlyacc: rely on runtime/caml/version.h

Before this commit a dedicated version.h file was generated.

shindere committed:
Remove the reference to the VERSION file from tools/pre-commit-githook

More precisely, before this commit the AUTOCONF_FILES variable was listing
the VERSION file, which is no longer necessary. Nothing needs to be added because
the build-aux/* also matches build-aux/ocaml_version.m4 where the
version of OCaml is now defined.

This commit also increases the HOOK_VERSION variable from 5 to 6.

shindere committed:
Update comment in utils/config.mlp

Make it po point to the right place where the version of OCaml is
actually defined.

shindere committed:
Let manual/src/html_processing/src/common.ml be generated by configure

damiendoligez commented on build-aux/ocaml_version.m4:
```diff
@@ -0,0 +1,72 @@
+m4_define([OCAML__VERSION_MAJOR], [4])
+m4_define([OCAML__VERSION_MINOR], [14])
+m4_define([OCAML__VERSION_PATCHLEVEL], [0])
+# NOte that the OCAML__VERSION_EXTRA string defined below is always empty
```
```suggestion
# Note that the OCAML__VERSION_EXTRA string defined below is always empty
```

damiendoligez commented on build-aux/ocaml_version.m4:
```diff
@@ -0,0 +1,72 @@
+m4_define([OCAML__VERSION],
+  [m4_do(
+    OCAML__VERSION_SHORT,
+    m4_if(OCAML__VERSION_PATCHLEVEL,[],[],.OCAML__VERSION_PATCHLEVEL),
```
You can get rid of this test, the patchlevel part is mandatory since 3.08.0.

damiendoligez commented on build-aux/ocaml_version.m4:
```diff
@@ -0,0 +1,72 @@
+# Other variants of the version needed here and there in the compiler
+
+m4_define([OCAML__VERSION_NUMERIC_PATCHLEVEL],
+  m4_if(OCAML__VERSION_PATCHLEVEL,[],[0],OCAML__VERSION_PATCHLEVEL))
```
Same here, patchlevel is always present.

damiendoligez reviewed (commented):
I can see why you don't want to compute the VERSION stuff at build time, but could you tell us why it has to be done at `autogen` time and not at `configure` time? It feels a bit odd to have to run autoconf whenever I want to change the version string.

gasche commented:
> There is also the case of tools/check-parser-uptodate-or-warn.sh, any
> thought on what should be done there, @gasche?

This tool checks whether the VERSION contains `+dev` and disables itself otherwise. (This was added after a bug in the script broke a user build.) One way to do this without a VERSION file would be to have relevant variables provided by `Makefile.config` -- similar to the variables of `version.h`, I guess. 

A propos: is there some documentation somewhere of the variables defined in `version.h`? I wonder, for example, what is the intended difference between OCAML_VERSION_ADDITIONAL and OCAML_VERSION_EXTRA.

shindere commented:
Damien Doligez (2021/09/21 06:58 -0700):
> @damiendoligez commented on this pull request.
>
> I can see why you don't want to compute the VERSION stuff at build
> time, but could you tell us why it has to be done at `autogen` time
> and not at `configure` time? It feels a bit odd to have to run
> autoconf whenever I want to change the version string.

How would you do to have it computed at configure time? Where would it
be stored, concretely?

> > +#*                                                                        *
> +#**************************************************************************
> +
> +# This file contains all the macros used to describe the current version of
> +# OCaml. It first defines the basic components and then computes all
> +# the different variants of the version used across the build system.
> +
> +# For the M4 macros defined below, we use the OCAML__VERSION (with a double
> +# underscore) to distinguish them from the C preprocessor macros which
> +# use a single underscore, since the two families of macros coexist
> +# in configure.ac.
> +
> +m4_define([OCAML__VERSION_MAJOR], [4])
> +m4_define([OCAML__VERSION_MINOR], [14])
> +m4_define([OCAML__VERSION_PATCHLEVEL], [0])
> +# NOte that the OCAML__VERSION_EXTRA string defined below is always empty
>
> ```suggestion
> # Note that the OCAML__VERSION_EXTRA string defined below is always empty
> ```
>
> > +# use a single underscore, since the two families of macros coexist
> +# in configure.ac.
> +
> +m4_define([OCAML__VERSION_MAJOR], [4])
> +m4_define([OCAML__VERSION_MINOR], [14])
> +m4_define([OCAML__VERSION_PATCHLEVEL], [0])
> +# NOte that the OCAML__VERSION_EXTRA string defined below is always empty
> +# for officially-released versions of OCaml.
> +m4_define([OCAML__VERSION_EXTRA], [dev0-2021-06-03])
> +m4_define([OCAML__VERSION_EXTRA_PREFIX], [+]) # could also be [~]
> +m4_define([OCAML__VERSION_SHORT], [OCAML__VERSION_MAJOR.OCAML__VERSION_MINOR])
> +# The OCAML__VERSION below must be in the format specified in stdlib/sys.mli
> +m4_define([OCAML__VERSION],
> +  [m4_do(
> +    OCAML__VERSION_SHORT,
> +    m4_if(OCAML__VERSION_PATCHLEVEL,[],[],.OCAML__VERSION_PATCHLEVEL),
>
> You can get rid of this test, the patchlevel part is mandatory since
> 3.08.0.

Ah, I didn't know. Well still, does it hurt to leave it, really?

> > +# still correct and this file can thus still be used to figure it out,
> +# the version itself is actually defined in the build-aux/ocaml_version.m4
> +# file (See the OCAML__VERSION* macros there.)
> +# To update the present VERSION file:
> +# 1. Update build-aux/ocaml_version.m4
> +# 2. Run tools/autogen.
> +# 3. If you are in a context where version control matters,
> +# commit the changes to both build-aux/ocaml_version.m4 and VERSION.
> +# The version string must be in the format described in stdlib/sys.mli
> +END_OF_VERSION_FILE
> +])
> +
> +# Other variants of the version needed here and there in the compiler
> +
> +m4_define([OCAML__VERSION_NUMERIC_PATCHLEVEL],
> +  m4_if(OCAML__VERSION_PATCHLEVEL,[],[0],OCAML__VERSION_PATCHLEVEL))
>
> Same here, patchlevel is always present.

Same answer / question from my side.

shindere commented:
Gabriel Scherer (2021/09/21 07:26 -0700):
> > There is also the case of tools/check-parser-uptodate-or-warn.sh, any
> > thought on what should be done there, @gasche?
>
> This tool checks whether the VERSION contains `+dev` and disables
> itself otherwise. (This was added after a bug in the script broke a
> user build.) One way to do this without a VERSION file would be to
> have relevant variables provided by `Makefile.config` -- similar to
> the variables of `version.h`, I guess.

I can definitely make the variables available to the build system. Just
that they won't be abvailable to the script itself. Currently I can see
two different ways to go about that. Either this script gets generated,
too, and we generate a no-op script for public releasess. Or we put the
test in a Makefile, i.e. outisde of this script. Do you have a
preference, @gashce?

> A propos: is there some documentation somewhere of the variables
> defined in `version.h`?

I tried to document them in `build-aux/ocaml_version.m4`.

> I wonder, for example, what is the intended difference between
> OCAML_VERSION_ADDITIONAL and OCAML_VERSION_EXTRA.

Thanks a lot for asking, @gasche. I have to admit: they are synonymous.
I introduced "extra" because I liked it better and it was shorter than
"additional". Myh intention was to figure out whether the additional
compnent is ever used in any opam package but I didn't take the time to
do so. It's not used in the compiler's codebase so we could simply
remove it completely. Of course I know than the more conservative (and
thus safe) solution would be to stay with additional and simply not add
the "extra" oone, so if people insist I'll stay with "additional"
although I don't like it too much and don't find it that nice.

gasche commented:
> Either this script gets generated,
> too, and we generate a no-op script for public releases. Or we put the
> test in a Makefile, i.e. outside of this script.

I see. I would prefer to stick with a shell script instead of weird Makefile logic, so I guess the first option would be nice. We could have a `tools/foo.sh.in` script with a line like `OCAML_VERSION="@OCAML_VERSION_STRING@"` and then grepping on that.

damiendoligez commented:
> How would you do to have it computed at configure time? Where would it be stored, concretely?

From VERSION, mostly using the same script that is currently used.

> Ah, I didn't know. Well still, does it hurt to leave it, really?

Not really. It's just a bit of useless code that makes the file slightly harder to read.

shindere commented:
Gabriel Scherer (2021/09/21 07:59 -0700):
> > Either this script gets generated,
> > too, and we generate a no-op script for public releases. Or we put the
> > test in a Makefile, i.e. outside of this script.
>
> I see. I would prefer to stick with a shell script instead of weird
> Makefile logic, so I guess the first option would be nice. We could
> have a `tools/foo.sh.in` script with a line like
> ***@***.***_VERSION_STRING@"` and then grepping on that.

That could be done, yes. But the analysis to figure out whether it's a
dev script could take place during configure, and then there wold need t
be no test at all in the generated script. Do you see what I mean?

shindere commented:
Damien Doligez (2021/09/21 08:00 -0700):
> > How would you do to have it computed at configure time? Where would it be stored, concretely?
>
> From VERSION, mostly using the same script that is currently used.

No, please no.

I find it awkward to start from big things and then have logic to split
them in smaller bits. To me it's way cleaner to start with smaller,
"atomic" parts and then build bigger pieces from that.

> > Ah, I didn't know. Well still, does it hurt to leave it, really?
>
> Not really. It's just a bit of useless code that makes the file
> slightly harder to read.

Well ultimately I can remove it if you like. I think I did that because
that kind of code was also present in the script used to generate
version.h.

gasche commented:
I see what you mean, but I suspect that this approach will make the plumbing more complex and harder to understand and modify for people not familiar with the configure script.

shindere commented:
Gabriel Scherer (2021/09/21 08:31 -0700):
> I see what you mean, but I suspect that this approach will make the
> plumbing more complex and harder to understand and modify for people
> not familiar with the configure script.

I think things will be clearer when there is an iimplemntation to
discuss on. Will come up with that at the beginning of next week (away
from tonight till then).

dra27 commented:
I did some similar work to this as part of my "no-scripting" work, however my take was slightly different in that I eliminated `make-version-header.sh` by folding the functionality into sak (so still generating the header at build time). For that, it was logical to remove `VERSION` as @shindere suggests but I made the version string passed to `AC_INIT` in `configure.ac` the main source and put that string (which autoconf provides in `@PACKAGE_VERSION@`) into `Makefile.config`. For `tools/check-parser-uptodate-or-warn.sh`, it's much easier to control the check in the `Makefile` (i.e. not call the script at all if not in developer mode) which I did in https://github.com/dra27/ocaml/pull/35/commits/3d9f36cafa5a187e28b387768df2ac5bb221702b.

Regarding `Makefile` magic versus shell magic, the former is portable (I mean actually portable, not just that it runs on POSIX) and the latter is not - FWIW, I'm at presently quite strongly arguing that we excise as much of the scripting as possible from the build system! I'm sure we can find a setting of the `LC_` variables where grepping for `+dev` fails because the meaning of something was changed...

@damiendoligez - if you update the version string today, you should still run `tools/autogen` (or `make configure`), you just end up with the `configure` script and `VERSION` file out of sync (CI ensures this would never get pushed). Both my approach and @shindere's "force" you to do the process properly if changing the version string. Note that we have `make reconfigure`, so it'd be possible for this to be automated if we wanted (I'm not necessarily sure that we do, though) - i.e. automatically run `make reconfigure` if `configure.ac` or the auxiliaries is newer than `Makefile.config`.

Finally, I think @shindere's approach of generating these files in `configure` is superior to mine doing it via `runtime/sak`.

gasche commented:
> For tools/check-parser-uptodate-or-warn.sh, it's much easier to control the check in the Makefile (i.e. not call the script at all if not in developer mode) which I did in dra27@3d9f36c.

Ah yes, this is a nice approach, thanks!

shindere commented:
Thanks David!

Will submit patches taking the comments into account from next Monday
onwards.

damiendoligez commented:
> No, please no. I find it awkward to start from big things and then have logic to split them in smaller bits. To me it's way cleaner to start with smaller, "atomic" parts and then build bigger pieces from that.

Ok, I see what you mean.

> Well ultimately I can remove it if you like. I think I did that because that kind of code was also present in the script used to generate version.h.

Please yes. That script has the test because it is very old, not because the test is useful.

shindere commented:
One other question before I restart working on this.

Does somebody think it would be a wrong idea to expose the different bits
of the version number in the `Sys` module, besides to the
`ocaml_version` string? For instance we could have:

```
val ocaml_version_major : int
val ocaml_version_minor : int
val ocaml_version_patch_level : int
val ocaml_version_extra : string
val ocaml_version_extra_prefix : char
```

If nobody disagrees that's something I'd joyfully add to the PR.
The machinery is already in place so it wouldn't be a big work.

dra27 commented:
It would (one day) allow code like this to disappear (taken from [Dune's bootstrap](https://github.com/ocaml/dune/blob/main/bootstrap.ml#L68)):

```ocaml
let v = Scanf.sscanf Sys.ocaml_version "%d.%d.%d" (fun a b c -> (a, b, c))
```

I wouldn't expose `caml_version_extra_prefix`, though. If a user has some mildly strange reason to care about exactly whether `+`, `#`, they can look at the string!

Perhaps using a tuple/record `ocaml_release` containing major, minor and patch-level - pattern matching is easy and polymorphic compare does what's wanted for comparing versions?

shindere commented:
Thanks.

I thought about a record yes but was unsure aobut hte name.

Any other opinion would be really appreciated. Otherwise I guess I'd
tend to follow David suggestion, my personal preference being for a
record over a tuple. So the name of the record would be `ocaml_release`
with the four fields `major`, `minor`, `patchlevel` and `extra`, the
three first being of type `int` and the last one being of tye `string`.

dra27 commented:
I personally wouldn't put the extra part in the record - the useful thing is to be able to write `Sys.ocaml_release >= (4, 13, 0)`, for example, without having to worry about it was "beta1".

shindere commented:
David Allsopp (2021/09/24 04:33 -0700):
> I personally wouldn't put the extra part in the record - the useful
> thing is to be able to write `Sys.ocaml_release >= (4, 13, 0)`, for
> example, without having to worry about it was "beta1".

OK and then you'd prefer a tuple?

shindere commented:
And owuld you publish the extra bit, or not at all? How would you call
it? It would feel odd to call it _ocaml_version_extra` because the
`ocaml_version` string does already contain the `extra` (or additional)
bit.

shindere commented:
David Allsopp (2021/09/24 04:33 -0700):
> I personally wouldn't put the extra part in the record - the useful
> thing is to be able to write `Sys.ocaml_release >= (4, 13, 0)`, for
> example, without having to worry about it was "beta1".

Well would it really change anything to add the extra bit? As long as it
appears at the end, you could add an empty string when comparing, no?

shindere commented:
I don't know but confusely I feel that not exposing everything would in
a way be synonymous of stopping halfway, not doing the job completely.
After all, what does it cost to us to expose everything, in whatever
way, even if nobody makes use of it? On the other side, if we don't
expose everything now and the use appears one day to access something
that had not been exposed so far, then we will have to add it and users
will have to discriminate between the versions where not everything is
exposed and those where everything is exposed so it feels to me if we do
it now, we should do it as cleanly, as completely and in an as stable
way as possible to avoid having to change it again in the future in any
way, to the best we can.

damiendoligez commented:
>I feel that not exposing everything would in
a way be synonymous of stopping halfway, not doing the job completely.

Same here, but... if you add the "extra" string, polymorphic comparison doesn't quite do the right thing, because the canonical order on version numbers is a modified lexicographic order where "~" is lower than the empty string.

So we should probably add the record with all the info, and someday we'll add the comparison function (maybe).

shindere commented:
Damien Doligez (2021/09/29 05:40 -0700):
> >I feel that not exposing everything would in
> a way be synonymous of stopping halfway, not doing the job completely.
>
> Same here, but... if you add the "extra" string, polymorphic comparison doesn't quite do the right thing, because the canonical order on version numbers is a modified lexicographic order where "~" is lower than the empty string.
>
> So we should probably add the record with all the info, and someday
> we'll add the comparison function (maybe).

That, or we add the extra bit in another place?

Would the situation be different with '-' rather than '~'?

shindere commented:
Damien Doligez (2021/09/21 06:58 -0700):
> @damiendoligez commented on this pull request.
>
> I can see why you don't want to compute the VERSION stuff at build
> time, but could you tell us why it has to be done at `autogen` time
> and not at `configure` time? It feels a bit odd to have to run
> autoconf whenever I want to change the version string.

I think @dra27 already addressed this point.

> > +#*                                                                        *
> +#**************************************************************************
> +
> +# This file contains all the macros used to describe the current version of
> +# OCaml. It first defines the basic components and then computes all
> +# the different variants of the version used across the build system.
> +
> +# For the M4 macros defined below, we use the OCAML__VERSION (with a double
> +# underscore) to distinguish them from the C preprocessor macros which
> +# use a single underscore, since the two families of macros coexist
> +# in configure.ac.
> +
> +m4_define([OCAML__VERSION_MAJOR], [4])
> +m4_define([OCAML__VERSION_MINOR], [14])
> +m4_define([OCAML__VERSION_PATCHLEVEL], [0])
> +# NOte that the OCAML__VERSION_EXTRA string defined below is always empty
>
> ```suggestion
> # Note that the OCAML__VERSION_EXTRA string defined below is always empty
> ```

Fixed. Thanks.

> > +# use a single underscore, since the two families of macros coexist
> +# in configure.ac.
> +
> +m4_define([OCAML__VERSION_MAJOR], [4])
> +m4_define([OCAML__VERSION_MINOR], [14])
> +m4_define([OCAML__VERSION_PATCHLEVEL], [0])
> +# NOte that the OCAML__VERSION_EXTRA string defined below is always empty
> +# for officially-released versions of OCaml.
> +m4_define([OCAML__VERSION_EXTRA], [dev0-2021-06-03])
> +m4_define([OCAML__VERSION_EXTRA_PREFIX], [+]) # could also be [~]
> +m4_define([OCAML__VERSION_SHORT], [OCAML__VERSION_MAJOR.OCAML__VERSION_MINOR])
> +# The OCAML__VERSION below must be in the format specified in stdlib/sys.mli
> +m4_define([OCAML__VERSION],
> +  [m4_do(
> +    OCAML__VERSION_SHORT,
> +    m4_if(OCAML__VERSION_PATCHLEVEL,[],[],.OCAML__VERSION_PATCHLEVEL),
>
> You can get rid of this test, the patchlevel part is mandatory since
> 3.08.0.

I did that and documented that the patchlevel is mandatory, has to be a
number and should not be left-padded with zeros, same for major and
minor numbers.

>
> > +# still correct and this file can thus still be used to figure it out,
> +# the version itself is actually defined in the build-aux/ocaml_version.m4
> +# file (See the OCAML__VERSION* macros there.)
> +# To update the present VERSION file:
> +# 1. Update build-aux/ocaml_version.m4
> +# 2. Run tools/autogen.
> +# 3. If you are in a context where version control matters,
> +# commit the changes to both build-aux/ocaml_version.m4 and VERSION.
> +# The version string must be in the format described in stdlib/sys.mli
> +END_OF_VERSION_FILE
> +])
> +
> +# Other variants of the version needed here and there in the compiler
> +
> +m4_define([OCAML__VERSION_NUMERIC_PATCHLEVEL],
> +  m4_if(OCAML__VERSION_PATCHLEVEL,[],[0],OCAML__VERSION_PATCHLEVEL))
>
> Same here, patchlevel is always present.

I got rid of the `OCAML__VERSION_NUMERIC_PATCHLEVEL` macro completely.
The code looks more regular that way, which is cool.

damiendoligez commented:
>Would the situation be different with '-' rather than '~'?

I'm not sure I understand the question. The ordering of version numbers was (IIUC) invented by Debian and now the whole world of free software is standardizing on it, so it's not something that we can choose, and it's not the lexicographic order implemented by polymorphic equality.

But I don't think many programs will need to compare versions down to this level anyway, and an implementation of the order is probably already available in OPAM sources.

shindere commented:
OK, thanks.

shindere commented:
Gabriel Scherer (2021/09/22 01:07 -0700):
> > For tools/check-parser-uptodate-or-warn.sh, it's much easier to control the check in the Makefile (i.e. not call the script at all if not in developer mode) which I did in ***@***.***
>
> Ah yes, this is a nice approach, thanks!

So I just implemented that.

There are two differences between my implementation and the one done by
David.

First, I don't look for the '+', just for the 'dev' sub-string.

Second and perhaps more importantly, my implementation is more
conservative, in the sense that, if building a dev version,
`tools/check-parser-uptodate-or-warn.sh` is added to the dependencies of
`parsing/parser.ml`. The cost for that is that I have to duplicate a bit
of code, namely the call to the `findstring` function, but it felt
better to me to stick to the former dependencies.

In the ned, I am wondering what we shall do with the VERSION file?

It's not installed anyway so if nothing uses it in the build system, is
it actually worth even generating it? Perhaps package maintainers will
appreciate to have it kept?

shindere committed:
Move version check from tools/check-parser-uptodate-or-warn.sh to Makefile

This commit introduces the different components of an OCaml version
as build variables.

It thus becomes possible to test from within the build system whether
the currently built version of OCaml is a development version or not.

This ability is used to move the development version test from the
beginning of tools/check-parser-uptodate-or-warn.sh to the root Makefile,
which will make that test portable even to non-Posix systems and
no matter how the locales are configured.

shindere commented:
Damien Doligez (2021/09/22 06:39 -0700):
> > Well ultimately I can remove it if you like. I think I did that because that kind of code was also present in the script used to generate version.h.
>
> Please yes. That script has the test because it is very old, not
> because the test is useful.

So as mentionned above, the tests have now been removed.

lthls commented:
As long as you're changing how the version information is generated, would you mind making the difference between development and release versions explicit, with a dedicated macro ? There are at least three places where the build system matches on the '+dev' or 'dev' string in the version, and I think it's a bit fragile (plus I got annoyed that the Flambda repo had to put '+dev' into its version string to be usable as a development repo, but I guess this should not concern you).

shindere commented:
Sure, we can do that!

Where exactly do you need this, please?

Would you mind pointing to the places in the build system were +dev is
mentionned?

Actually, rather than matching on +dev, perhaps we should have a real
macro saying explicitly whether a version is a development one or not,
independently of the name. What do you think?

shindere committed:
Add a macro to explicitly control whether this is a dev version or not

shindere commented:
@lthls I just pushed an implementation of your suggestion.

Would you mind having a look to it and letting me know whether that suits your needs,please?

dra27 commented on configure.ac:
```diff
@@ -591,8 +610,8 @@ AS_CASE([$ocaml_cv_cc_vendor],
 
-AS_CASE([$enable_warn_error,AC_PACKAGE_VERSION],
-  [yes,*|,*+dev*],
+AS_CASE([$enable_warn_error,OCAML_DEVELOPMENT_VERSION],
```
This either needs to be `OCAML__DEVELOPMENT_VERSION` (two underscores) or `$OCAML_DEVELOPMENT_VERSION` (at least I think the `$` form works too - I haven't checked). The `diff` for configure shows that isn't expanding as it stands, as it's not an m4 macro.

dra27 commented on configure.ac:
```diff
@@ -1873,8 +1892,8 @@ AS_IF([test "x$documentation_tool_cmd" = 'x']
 
-AS_CASE([$enable_ocamltest,AC_PACKAGE_VERSION],
-  [yes,*|,*+dev*],[ocamltest='ocamltest'],
+AS_CASE([$enable_ocamltest,OCAML_DEVELOPMENT_VERSION],
```
Same problem here

dra27 commented on runtime/Makefile:
```diff
@@ -37,8 +37,8 @@ NATIVE_C_SOURCES := $(addsuffix .c, \
 
-GENERATED_HEADERS := caml/opnames.h caml/version.h caml/jumptbl.h build_config.h
-CONFIG_HEADERS := caml/m.h caml/s.h
+GENERATED_HEADERS := caml/opnames.h caml/jumptbl.h build_config.h
+CONFIGURED_HEADERS := caml/m.h caml/s.h caml/version.h
```
It took me a second or two to realise what was changing here. It's very minor, but might it be worth a comment above: `# $(GENERATED_HEADERS) are built by make; $(CONFIGURED_HEADERS) are built by configure`?

lthls commented:
Yes, that's what I had in mind. Thanks !

shindere committed:
Make the documentation of ocaml_version in stdlib/sys.mli more precise

dra27 reviewed (commented):
I haven't double-checked it, but I don't think the `+dev` change is working (see comments). I do think it's a good idea, though!

For what it's worth, I'm not sure that the code would be much worse for having the version string provided in `AC_INIT` and then parse/validate it using macros in `ocaml_version.m4` - it takes a lot of reading in `ocaml_version.m4` to check what the version actually is

shindere commented:
Vincent Laviron (2021/09/29 15:36 +0000):

My pleausre!

Do you see other places in the build system where this could be used?

shindere commented:
David Allsopp (2021/09/29 15:42 +0000):
> @dra27 commented on this pull request.
>
> I haven't double-checked it, but I don't think the `+dev` change is
> working (see comments). I do think it's a good idea, though!

You are right David! Many thanks for having paid attention to this!

> For what it's worth, I'm not sure that the code would be much worse
> for having the version string provided in `AC_INIT` and then
> parse/validate it using macros in `ocaml_version.m4` - it takes a lot
> of reading in `ocaml_version.m4` to check what the version actually is

Hmm, sorry about that, but after all this work I would prefer to leave
things as they are on that front. I assume it's also a matter of
habbits, and if it's jsut a matter of figuring out the version quickly
then, well, we can always keep the VERSION file just for that.


>
> > @@ -591,8 +610,8 @@ AS_CASE([$ocaml_cv_cc_vendor],
>    warn_error_flag='-Werror'
>    cc_warnings='-Wall -Wdeclaration-after-statement'])
>
> -AS_CASE([$enable_warn_error,AC_PACKAGE_VERSION],
> -  [yes,*|,*+dev*],
> +AS_CASE([$enable_warn_error,OCAML_DEVELOPMENT_VERSION],
>
> This either needs to be `OCAML__DEVELOPMENT_VERSION` (two underscores) or `$OCAML_DEVELOPMENT_VERSION` (at least I think the `$` form works too - I haven't checked). The `diff` for configure shows that isn't expanding as it stands, as it's not an m4 macro.
>
> > @@ -1873,8 +1892,8 @@ AS_IF([test "x$documentation_tool_cmd" = 'x']
>
>
>
> -AS_CASE([$enable_ocamltest,AC_PACKAGE_VERSION],
> -  [yes,*|,*+dev*],[ocamltest='ocamltest'],
> +AS_CASE([$enable_ocamltest,OCAML_DEVELOPMENT_VERSION],
>
> Same problem here

Fixed, thanks. I used the macros because I think it's more efficient,
since they wil be evaluated by autoconf rather than by configure.

> > +GENERATED_HEADERS := caml/opnames.h caml/jumptbl.h build_config.h
> +CONFIGURED_HEADERS := caml/m.h caml/s.h caml/version.h
>
> It took me a second or two to realise what was changing here. It's
> very minor, but might it be worth a comment above: `#
> $(GENERATED_HEADERS) are built by make; $(CONFIGURED_HEADERS) are
> built by configure`?

Well am I wrong that these variables were already there and that the
only that has actually been introduced by this PR is to move
`caml/version.h` from one variable to the other and renaming
`CONFIG_HEADERS` to `CONFIGURED_HEADERS`? I followed your suggestion
anyway, thanks. I hope things are clearer now?

lthls commented:
> My pleausre! Do you see other places in the build system where this could be used?

You've found all the ones I knew about. Eventually we could consider extending the mechanism to other places (maybe `ocamldoc` and the manual builds ?), but that doesn't need to be discussed in this PR.

shindere commented:
Vincent Laviron (2021/09/29 09:24 -0700):
> > My pleausre! Do you see other places in the build system where this could be used?
>
> You've found all the ones I knew about. Eventually we could consider
> extending the mechanism to other places (maybe `ocamldoc` and the
> manual builds ?), but that doesn't need to be discussed in this PR.

Yeah it feels a bit ambitious for me, not even sure I'm the most
qualified to implement these extensions. But many thanks for all your
suggestions, I like them and find them very cool.

shindere commented:
I just pushed a commit that exposes the version-related information in the
Sys module of the standard library.

The way it's done slightly differs from what has been discussed so far
and might be found a bit convoluted, but I wanted to propose this
implementation because that's the one I like. More objectively, it pursues
two purposes:

 1. Making all the information available
 2. Using our type-system to make it available in a clear way.

Some documentation is still missing. I was not sure whether it should be
put close to the types, or close to the values, or whether perhaps both
should be documented?

shindere commented:
Gabriel Scherer (2021/09/21 07:26 -0700):
> A propos: is there some documentation somewhere of the variables
> defined in `version.h`? I wonder, for example, what is the intended
> difference between OCAML_VERSION_ADDITIONAL and OCAML_VERSION_EXTRA.

All the variables are documented in `build-aux/ocaml_version.m4`.

As I said, my plan is to get rid of OCAML_VERSION_ADDITIONAL and to have
it replaced by OCAML_VERSION_EXTRA in the few places where it's used,
namely nowhere in the compiler code-base and only in the bsbnative and
mirage packages otherwise, according to `opam grep`.

shindere commented:
This has just been rebased on latest trunk.

Gentle ping to reviewers so that this can finally be approved/merged.

Thanks!

shindere commented:
Could someone copy/paste the error messages produced by CI, please?

dra27 commented on VERSION:
```diff
@@ -1,4 +1,12 @@
+# 2. Run tools/autogen.
+# 3. If you are in a context where version control matters,
+# commit the changes to both build-aux/ocaml_version.m4 and VERSION.
+# The version string must be in the format described in stdlib/sys.mli
```
`VERSION` now has more than 9 lines, so `check-typo` expects it to have a valid copyright header. `typo.missing-header` needs to go in `.gitattributes`, therefore.

dra27 commented on .gitignore:
```diff
@@ -97,6 +97,7 @@ _build
 /lex/parser.output
 
 /manual/src/cmds/warnings-help.etex
+/manual/src/html_processing/src/common.ml
```
This needs to be removed by the `distclean` target in `manual/src/html_processing/Makefile`

dra27 commented on manual/src/Makefile:
```diff
@@ -122,9 +122,6 @@ etex-files: $(FILES)
 	$(TEXQUOTE) < $< > $*.texquote_error.tex
 	mv $*.texquote_error.tex $@
 
-version.tex: $(ROOTDIR)/VERSION
-	sed -n -e '1s/^\([0-9]*\.[0-9]*\).*$$/\\def\\ocamlversion{\1}/p' $< > $@
-
```
`version.tex` is included in the `FILES` variable in `manual/src/Makefile` which means it gets removed by the clean target. Something needs to happen to so that `distclean` removes `version.tex` instead.

dra27 reviewed (commented):
CI issues highlighted in-place

shindere commented:
Thank you so much for the hints, David!

I'll process your feedback ASAP.

I am wondering what I could/shuld do to be more autonomous about the CI
feedbacks, it's quite frustrating to have to depend on others to figure
out what's going wrong, inspite of the kindness and support of all the
community.

dra27 commented:
@shindere - in this case, the logs are truly dreadful (I wrote them, so I'll take the blame for that)! Some of the actual errors of some hygiene jobs are hidden a long way up in the output, so I checked exactly why each failure was occurring as a penance for this! The reason it's ended up this way is that rather than stopping on the first problem, the CI now carries on to do as much testing as still makes sense. The benefit is that you don't, for example, fix check-typo things and only then discover that `configure` is stale, for example.

We should alter the GHA runs to upload their logs as artefacts (especially the main build log) so that they can at least be downloaded from the API. I expect it'll be hard to get to that this month, but hopefully next I can set some time aside for that.

shindere commented:
No problem, I thought it had more to do with the graphical interface.

Ideally, if a failure occurs, I'd like thelogs to be sent to me by
e-mail. Do you think that'd be possible?

damiendoligez commented on build-aux/ocaml_version.m4:
```diff
@@ -0,0 +1,91 @@
+#*                                                                        *
+#*                                 OCaml                                  *
+#*                                                                        *
+#*            Sebastien Hinderer, projet Gallium, INRIA Paris             *
```
Projet Gallium doesn't exist anymore...

```suggestion
#*            Sebastien Hinderer, projet Cambium, INRIA Paris             *
```

damiendoligez commented on build-aux/ocaml_version.m4:
```diff
@@ -0,0 +1,91 @@
+# by OCAML__VERSION_MAJOR.OCAML__VERSION_MINOR.OCAML__VERSION_PATCHLEVEL
+# and with an empty OCAML__VERSION_EXTRA string.
+
+m4_define([OCAML__VERSION_EXTRA_PREFIX], [+])
```
According to your text above, this + should be a ~
```suggestion
m4_define([OCAML__VERSION_EXTRA_PREFIX], [~])
```

damiendoligez commented on stdlib/sys.ml.in:
```diff
@@ -127,6 +127,26 @@ external runtime_warnings_enabled: unit -> bool =
 
+let development_version = @OCAML_DEVELOPMENT_VERSION@
+
+type extra_prefix = Plus | Tilde
+
+type extra_info = extra_prefix * string
```
You could also write:
```
type extra_info =
| Plus of string
| Tilde of string
| Empty
```

damiendoligez commented on stdlib/sys.mli:
```diff
@@ -337,9 +337,27 @@ val ocaml_version : string
     [additional-info] is an arbitrary string.
-    The [[.patchlevel]] part is absent for versions anterior to 3.08.0.
+    The [[.patchlevel]] part was absent before version 3.08.0 and
+    became mandatory from 3.08.0 (included) onwards.
```
The `(included)` part is redundant.
```suggestion
    became mandatory from 3.08.0 onwards.
```

damiendoligez reviewed (approved):
Looks good. Here are a few suggestions that you might choose to follow.

shindere commented:
I think I have fixed everything, many thanks for the detailed report,
@dra27.

Let's see what CI thinks about all this and if it is happy I will merge.

shindere committed:
Define OCaml's version in build-aux/ocaml_version.m4

Before this commit, the version of OCaml was defined on the first line
of the VERSION file present at the top-level of the source tree.

This commit defines the version of OCaml in build-aux/ocaml_version.m4
as a series of macros defining each component of the version and
then putting all these components together in the right way to
obtain the main version string and the other variants used here and there.

With this commit, the VERSION file becomes a by-product of running autoocnf,
which is usually done by running tools/autogen.

However, this commit takes special care of generating the very same
VERSION file than the one used before the commit. The modifications
in the comments will be introduced in subsequent commits, for the
sake of clarity and to make it possible and eaiser to verify that
nothing has been broken or altered.

With this commit all the components of the version number become available
to the configure script which will thus be available to propagate them
to the whole build system.

shindere committed:
Add detailed information about the current version of OCaml to the Sys module

shindere commented:
Damien Doligez (2021/10/06 03:31 -0700):
> @damiendoligez approved this pull request.
>
> Looks good. Here are a few suggestions that you might choose to follow.
>
> > @@ -0,0 +1,91 @@
> +#**************************************************************************
> +#*                                                                        *
> +#*                                 OCaml                                  *
> +#*                                                                        *
> +#*            Sebastien Hinderer, projet Gallium, INRIA Paris             *
>
> Projet Gallium doesn't exist anymore...
>
> ```suggestion
> #*            Sebastien Hinderer, projet Cambium, INRIA Paris             *
> ```

Fixed, thanks.

> > +m4_define([OCAML__VERSION_MAJOR], [4])
> +m4_define([OCAML__VERSION_MINOR], [14])
> +m4_define([OCAML__VERSION_PATCHLEVEL], [0])
> +# Note that the OCAML__VERSION_EXTRA string defined below is always empty
> +# for officially-released versions of OCaml.
> +m4_define([OCAML__VERSION_EXTRA], [dev0-2021-06-03])
> +# The OCAML__VERSION_EXTRA_PREFIX macro defined below should be a
> +# single character:
> +# Either [~] if this version is published before the version represented
> +# by OCAML__VERSION_MAJOR.OCAML__VERSION_MINOR.OCAML__VERSION_PATCHLEVEL
> +# and with an empty OCAML__VERSION_EXTRA string;
> +# Or [+] if this version is published after the version represented
> +# by OCAML__VERSION_MAJOR.OCAML__VERSION_MINOR.OCAML__VERSION_PATCHLEVEL
> +# and with an empty OCAML__VERSION_EXTRA string.
> +
> +m4_define([OCAML__VERSION_EXTRA_PREFIX], [+])
>
> According to your text above, this + should be a ~
> ```suggestion
> m4_define([OCAML__VERSION_EXTRA_PREFIX], [~])
> ```

I discovered later that it was the text which was wrong so I fixed the
text rather than the code. Thanks.

> > +type extra_prefix = Plus | Tilde
> +
> +type extra_info = extra_prefix * string
>
> You could also write:
> ```
> type extra_info =
> | Plus of string
> | Tilde of string
> | Empty
> ```

Indeed. I understand that this flattened flavour would be slightly more
efficient and less memory-consuming but I prefere the current one
because the `string` bit (1) does not need to be repeated and (2) is
enforced for both constructors at once.

> > @@ -337,9 +337,27 @@ val ocaml_version : string
>        ["major.minor[.patchlevel][(+|~)additional-info]"],
>      where [major], [minor], and [patchlevel] are integers, and
>      [additional-info] is an arbitrary string.
> -    The [[.patchlevel]] part is absent for versions anterior to 3.08.0.
> +    The [[.patchlevel]] part was absent before version 3.08.0 and
> +    became mandatory from 3.08.0 (included) onwards.
>
> The `(included)` part is redundant.
> ```suggestion
>     became mandatory from 3.08.0 onwards.
> ```

Fixed.

shindere committed:
Let configure rather than make generate manual/src/version.tex

This also required a modification of the root makefile so that its
clean and distclean targets also call the corresponding ones in
the manual.
